libcaf_core/caf/flow/step/take_last.hpp

```diff
@@ -0,0 +1,55 @@
+// This file is part of CAF, the C++ Actor Framework. See the file LICENSE in
+// the main distribution directory for license terms and copyright or visit
+// https://github.com/actor-framework/actor-framework/blob/master/LICENSE.
+
+#pragma once
+
+#include "caf/detail/ring_buffer.hpp"
+#include "caf/fwd.hpp"
+
+#include <cstddef>
+
+namespace caf::flow::step {
+
+template <class T>
+class take_last {
+public:
+  using input_type = T;
+
+  using output_type = T;
+
+  explicit take_last(size_t num) : elements_(num) {
+    // nop
+  }
+
+  take_last(take_last&&) = default;
+  take_last(const take_last&) = default;
+  take_last& operator=(take_last&&) = default;
+  take_last& operator=(const take_last&) = default;
+
+  template <class Next, class... Steps>
+  bool on_next(const input_type& item, Next& next, Steps&... steps) {
+    elements_.push_back(item);
+    return true;
+  }
+
+  template <class Next, class... Steps>
+  void on_complete(Next& next, Steps&... steps) {
+    while (!elements_.empty()) {
+      if (!next.on_next(elements_.front(), steps...))
+        break;
```

Comment:
This is a bug: we must not call *anything* on `next` if it returns `false`:

```suggestion
        return;
```